pages/api/VolunteerRoleService.ts

```diff
@@ -0,0 +1,187 @@
+/*
+ * @2023 Digital Aid Seattle
+ */
+
+import Airtable from 'airtable'
+import { DASVolunteerRole, DASVolunteerRoleBasicInfo } from 'types'
+
+// Documentation: Airtable provides an excellent guide to using their API.
+// To access documentation for our unique base, visit: https://airtable.com/{{baseID}}/api/docs#javascript/table:roles
+// To find the baseID, visit any page in the Digital Aid Airtable base and look at the URL.
+// It will be the string of characters after the first slash, starting with "app".
+
+
+class DASVolunteerRoleService {
+  private airtable: Airtable
+  private base: Airtable.Base
+  private volunteerRolesTable: Airtable.Table<Airtable.FieldSet>
+  private toolsWeUseTable: Airtable.Table<Airtable.FieldSet>
+
+  constructor(apiKey: string, baseId: string, tableId: string) {
+    this.airtable = new Airtable({
+      apiKey: apiKey || process.env.NEXT_PUBLIC_AIRTABLE_API_KEY,
+    })
+    this.base = this.airtable.base(
+      baseId || process.env.NEXT_PUBLIC_AIRTABLE_BASE_ID_DAS
+    )
+    this.volunteerRolesTable = this.base(
+      tableId || process.env.NEXT_PUBLIC_AIRTABLE_TABLE_VOLUNTEER_ROLES
+    )
+    this.toolsWeUseTable = this.base(process.env.NEXT_PUBLIC_AIRTABLE_TABLE_TOOLS_WE_USE)
+  }
+
+  async getAllActiveRoles(): Promise<DASVolunteerRoleBasicInfo[]> {
+    try {
+      const records = await this.volunteerRolesTable
+        .select({
+          view: 'Grid view',
+        })
+        .all()
+
+      const activeRoles = records.filter(
+        (record) => record.get('Status') === 'Active'
+      )
+
+      const volunteerRolesData = activeRoles.map((record) => {
+        return {
+          id: record.id,
+          role: String(record.fields.Role),
+          key: String(record.fields.Key),
+        }
+      })
+
+      return volunteerRolesData
+    } catch (err) {
+      console.log(err)
+      throw err
+    }
+  }
+  async getRoleDetailsByName(
+    roleKey: string
+  ): Promise<DASVolunteerRole | null> {
+    try {
+      const recordData = await this.volunteerRolesTable
+        .select({
+          view: 'Grid view',
+          maxRecords: 1,
+          filterByFormula: `{Key} = '${roleKey}'`,
+        })
+        .firstPage()
+      if (recordData.length === 0) {
+        return null
+      }
+      const record = recordData[0]
+
+      if (record.fields.Status !== 'Active') {
+        return null
+      }
+      let keyTechnologies = null
+
+      if (record.fields['Key technologies'] && Array.isArray(record.fields['Key technologies'])) {
```

Comment:
Lines 80-113 should be pushed down into a function.  One day we will write test methods <shock /> and smaller functions are easier to test.  It will also make the higher level function more intentional.

We may also find that we can cache the keyTechnologies.